iree/compiler/InputConversion/Common/QuantizedMatmulToMatmul.cpp

```diff
@@ -0,0 +1,236 @@
+          loc, zeroAcc.getType(), ValueRange{matrix}, ValueRange{zeroAcc},
+          indexingMaps, iterators,
+          [=](OpBuilder &b, Location loc, ValueRange args) {
+            Value matrixEl = args[0];
+            // Sign-extend the input matrix elem to accElTy before adding.
+            Value promotedMatrixEl =
+                b.create<arith::ExtSIOp>(loc, accElTy, matrixEl);
+            Value accEl = args[1];
+            Value sum = b.create<arith::AddIOp>(loc, promotedMatrixEl, accEl);
+            b.create<linalg::YieldOp>(loc, sum);
+          })
+      .getResult(0);
+}
+
+bool isConstantZero(Value val) {
+  auto constIntOp = val.getDefiningOp<arith::ConstantIntOp>();
+  return constIntOp && constIntOp.value() == 0;
+}
+
+// Pattern lowering quantized_matmul to matmul.
+// Always succeeds.
+//
+// This is implementing the math explained in Section 2.3 of
+// https://arxiv.org/abs/1712.05877.
+struct QuantizedMatmulToMatmul
+    : public OpRewritePattern<linalg::QuantizedMatmulOp> {
+  using OpRewritePattern<linalg::QuantizedMatmulOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::QuantizedMatmulOp quantizedMatmulOp,
+                                PatternRewriter &rewriter) const override {
+    Location loc = quantizedMatmulOp.getLoc();
+    ValueRange inputs = quantizedMatmulOp.inputs();
+    assert(inputs.size() == 4);
+    Value lhs = inputs[0];
+    Value rhs = inputs[1];
+    Value lhsZp = inputs[2];
+    Value rhsZp = inputs[3];
+    ValueRange outputs = quantizedMatmulOp.outputs();
+    // Compute the matmul part.
+    Value acc = outputs[0];
+    Value matmul = rewriter
+                       .create<linalg::MatmulOp>(loc, ValueRange{lhs, rhs},
+                                                 ValueRange{acc})
+                       .getResult(0);
+    bool lhsZpIsConstantZero = isConstantZero(lhsZp);
+    bool rhsZpIsConstantZero = isConstantZero(rhsZp);
+    if (lhsZpIsConstantZero && rhsZpIsConstantZero) {
+      // Easy case: both zero points are constant zeros, so the quantized_matmul
+      // was just a matmul all along.
+      rewriter.replaceOp(quantizedMatmulOp, matmul);
+      return success();
+    }
+    // Create the result. No need to zero-fill it as we will overwrite it.
+    ShapedType accType = acc.getType().cast<ShapedType>();
+    auto accDynShape = linalg::getDynOperands(loc, acc, rewriter);
+    Value initResult = rewriter.create<linalg::InitTensorOp>(
+        loc, accDynShape, accType.getShape(), accType.getElementType());
+    // Create the indexing maps for the generic.
+    MLIRContext *context = rewriter.getContext();
+    AffineExpr m, n;
+    bindDims(context, m, n);
+    AffineMap mapToNone = AffineMap::get(2, 0, context);
+    AffineMap mapToRowDim = AffineMap::get(2, 0, m, context);
+    AffineMap mapToColumnDim = AffineMap::get(2, 0, n, context);
+    AffineMap mapIdentity =
+        AffineMap::get(2, 0, ArrayRef<AffineExpr>{m, n}, context);
```

Comment:
nit: I think you can drop the `ArrayRef<AffineExpr>`...

Reply:
Thanks!